// @bp/nodes/core/common/lib/codeEditor/worker-bp.js
ace.define("ace/mode/bp_worker", ["require", "exports", "module", "ace/lib/oop", "ace/worker/mirror"], function(require, exports, module) {
  "use strict";

  var oop = require("ace/lib/oop");
  var Mirror = require("ace/worker/mirror").Mirror;

  var Worker = function(sender) {
    Mirror.call(this, sender);
    this.setTimeout(200);
    this.context = null;

    // Additional globals
    this.$globalVariables = ["ctx", "bp"];
  };

  oop.inherits(Worker, Mirror);

  (function() {
    this.onUpdate = function() {
      var value = this.doc.getValue();
      // Perform linting or other tasks here

      // Example: Simple error checking
      var errors = [];
      var lines = value.split("\n");
      for (var i = 0; i < lines.length; i++) {
        var line = lines[i];
        if (line.includes("undefinedVar")) {
          errors.push({
            row: i,
            column: line.indexOf("undefinedVar"),
            text: "Undefined variable: 'undefinedVar'",
            type: "error"
          });
        }
      }
      this.sender.emit("annotate", errors);
    };
  }).call(Worker.prototype);

  exports.Worker = Worker;
});
